### ISSUE TITLE
Bolt should use a configuration inheritance model similar to the one used by 'git'

### ISSUE THREAD

**Issue by trevor-vaughan** [puppetlabs/bolt#608]
When running `bolt` I expected the configuration options to be stacked in a method similar to those of `git` since some options are repository specific and some are user specific and should not be in a shared repository.

I did not expect the local configuration to override my global configuration completely (though it is documented at https://puppet.com/docs/bolt/0.x/configuring_bolt.html and I simply missed it).

Config inheritance order (from lowest to highest)
On non-Windows systems, the default config paths searched are:

1. `/etc/puppetlabs/bolt/bolt.yaml`
2. `~/.puppetlabs/etc/bolt/bolt.yaml`
3. `<boltdir>/bolt.yaml`

On Windows systems, the default config paths searched are:

1. `C:\Program Files\Puppet Labs\Bolt\bolt.yaml`
2. `%USERPROFILE%/.puppetlabs/etc/bolt/bolt.yaml`
3. `<boltdir>/bolt.yaml`

TODO:
- [x] Verify where Puppet looks for equivalent files on Windows
   ```
   Global: C:\ProgramData\PuppetLabs\bolt\etc
   Local: C:\Users\<user>\.puppetlabs\etc\bolt
   ```
- [x] Spike on how to merge configs. Any config options that's a hash should be evaluated for how it should be merged, and an RFC should be proposed to the Bolt team.
   https://gist.github.com/beechtom/f7b758893796d61a23c8ab868cf8b062
- [x] This should include a debug log message about what configs are being loaded and override precedence.

**Comment by lucywyman** [comment]
@trevor-vaughan  I think Alex has already talked to you about this, but just for visibility: There are some other features we'd like to add around this before implementing it so that users aren't blindsided by having configuration applied that they aren't sure where it's coming from (like a global config).  We'll leave this open since long term we plan to address it, but we don't have immediate plans or fixing. 

https://tickets.puppetlabs.com/browse/BOLT-795

**Comment by reidmv** [comment]
@lucywyman fwiw, I have the same request as @trevor-vaughan. I want to be able to commit a bolt.yaml to my e.g. control-repo in Puppet, to capture project-level configuration settings. However, I can't commit user-level configuration for ssh, winrm, and so forth to this file.

I need configuration stacking in order to use Bolt on a team level rather than an individual level.

**Comment by op-ct** [comment]
👍

The winner-takes-all nature of `bolt.yaml` files seems unusually restrictive when both users and projects can have good reasons to define their own settings.  This behavior will seem even stranger if system configuration files are implemented for #759.

As things stand, it doesn't make much sense to invest important settings in `~/.puppetlabs/bolt/bolt.yaml`, because none of them will work from any project directory—even if the only thing the directory's settings do is rename `site-modules/`.

FWIW, there have recently been similar requests in the puppetcommunity Slack:
* https://puppetcommunity.slack.com/archives/C7GP57T50/p1572033515064900
* https://app.slack.com/client/T0AQJ2CTU/C7GP57T50/thread/C7GP57T50-1571323080.293400
* https://puppetcommunity.slack.com/archives/C7GP57T50/p1571212261234100
